File: Numeros Pseudo-Aleatorios/XorShift 32, 64 e 128 bits/32 bits/xorshift32.py
"""
  PRNG .......: XorShift (64 bits)
  Ano ........: 2003
  Autor ......: George Marsaglia
"""

import time

class xorShift32State:
    def __init__(self):
        self.seed = 0

def xorshift32(state):
    random = state.seed & 0xFFFFFFFF # Máscara 32 bits
    random ^= random << 13
    random ^= random >> 17
    random ^= random << 5
    state.seed = random & 0xFFFFFFFF  

def main():
	
    Xn1 = xorShift32State()
    Xn2 = xorShift32State()

    Xn1.seed = 1234
    Xn2.seed = int(time.time())

    print("GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)")
    print("PSEUDORANDOM NUMBER GENERATOR (PRNG)")
    print("PADRAO: XorShift (32 bits)")
    print("=============================================")
    print()
    print("      Semente: 1234       Semente: TIME")
    print("      -------------       -------------")

    for i in range(1, 11):
        print("{:19d}{:20d}".format(xorshift32(Xn1), xorshift32(Xn2) & 0xFFFFFFFF))  # Máscara para garantir 32 bits

if __name__ == "__main__":
    main()

File: Numeros Pseudo-Aleatorios/XorShift 32, 64 e 128 bits/32 bits/xorshift32.c
/*
 PRNG .......: XorShift (32 bits)
 Ano ........: 2003
 Autor ......: George Marsaglia
*/

#include <stdio.h>
#include <stdint.h>
#include <time.h>

struct xorShift32State {
  uint32_t seed;
};

uint32_t xorshift32(struct xorShift32State *state)
{
  uint32_t random = state->seed;
  random ^= random << 13;
  random ^= random >> 17;
  random ^= random << 5;
  return state->seed = random;
}

int main(void)
{
	
  struct xorShift32State Xn1;
  struct xorShift32State Xn2;

  Xn1.seed = 1234;
  Xn2.seed = time(NULL);

  printf("GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)\n");
  printf("PSEUDORANDOM NUMBER GENERATOR (PRNG)\n");
  printf("PADRAO: XorShift (32 bits)\n");
  printf("=============================================\n");
  printf("\n");
  printf("      Semente: 1234       Semente: TIME\n");
  printf("      -------------       -------------\n");


  for (int i = 1; i <= 10; i++)
    printf("%19u%20u\n", xorshift32(&Xn1), xorshift32(&Xn2));

  return 0;
}
